### PR TITLE
[flutter_tool] --flutter_runner will download the debug symbols

### PR DESCRIPTION
**Pull request by iskakaushik**
This makes the flutter tool download the debug symbols for Flutter and Dart runners to be used in Fuchsia dev environments.

### CODE REVIEW

File: packages/flutter_tools/lib/src/cache.dart [hunk 1]
```diff
@@ -1058,6 +1059,36 @@ class FlutterRunnerSDKArtifacts extends CachedArtifact {
   }
 }
 
+/// The debug symbols for flutter runner for Fuchsia development.
+class FlutterRunnerDebugSymbols extends CachedArtifact {
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/cache.dart]: This needs a test - there should be some existing examples in cache.dart

Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/cache.dart]: cache_test.dart

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/cache.dart]: I've added a test. It doesn't actually test the downloading part, but tests the package resolution. I think this should cover any common failures.

File: packages/flutter_tools/lib/src/cache.dart [hunk 2]
```diff
@@ -1058,6 +1059,58 @@ class FlutterRunnerSDKArtifacts extends CachedArtifact {
   }
 }
 
+/// Implementations of this class can resolve URLs for packages that are versioned.
+///
+/// See also [CipdArchiveResolver].
+abstract class VersionedPackageResolver {
+  /// Returns the URL for the artifact.
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/cache.dart]: Give both of these classes a `const` constructor

File: packages/flutter_tools/lib/src/cache.dart [hunk 3]
```diff
@@ -1058,6 +1059,58 @@ class FlutterRunnerSDKArtifacts extends CachedArtifact {
...
+class CipdArchiveResolver extends VersionedPackageResolver {
+  @override
+  String resolveUrl(String packageName, String version) {
+    return '$_cipdBaseUrl/flutter/$packageName/+/git_revision:$version';
+  }
+}
+
+/// The debug symbols for flutter runner for Fuchsia development.
+class FlutterRunnerDebugSymbols extends CachedArtifact {
+  FlutterRunnerDebugSymbols(Cache cache, this.packageResolver, {this.dryRun = false})
```
Comment (jonahwilliams) [inline packages/flutter_tools/lib/src/cache.dart]: With the const constructor above you can make packageResolver a named argument that defaults to `const FlutterRunnerDebugSymbols`

Comment (iskakaushik) [reply packages/flutter_tools/lib/src/cache.dart]: Good idea. Done.

**Review by jonahwilliams — APPROVED** [review]
LGTM

**Pull request merged**